### ISSUE TITLE
Please add in

### ISSUE THREAD

**Issue by AndrewPardoe** [Microsoft/microsoft-pdb#1]
CVInfo is a good start, but please add information about the CVDump tool that outputs debug information from binaries in a more readable format.

**Comment by pombredanne** [comment]
:+1:

**Comment by assarbad** [comment]
:+1:

**Comment by nihilus** [comment]
It would be essential to have a nice parser like 'win_pdbx.zip' open-sourced for reading/writing pdb-files.
The header files themselves gives some hints on how to improve the state of win_pdbx.zip however.

**Comment by chwarr** [comment]
Looks like commit https://github.com/Microsoft/microsoft-pdb/commit/78aff39f352eab92e549a1e6c17d5c00f0d51e8c added a pre-compiled version of cvdump

**Comment by nihilus** [comment]
@chwarr That won't help me much for implementing PDB-support on Linux, will it?

**Comment by skochinsky** [comment]
@nihilus Why not? I hear Wine is pretty good at running win32 exes these days. Besides, most of cvdump source *is* in the repo, just a few pieces are missing so you can't build the same executable but can come pretty close. If you are having problems compiling things on Linux, why not open a new issue? @AndrewPardoe asked for the dumper and he got it, so I think *this* issue should be closed.

**Comment by pombredanne** [comment]
The source code is present in the repo afaik now. 
And this other tool by @jasonwhite is reportedd to be portable https://github.com/jasonwhite/ducible/tree/master/src/pdbdump

**Comment by pombredanne** [comment]
@JimRadigan "The source code is present in the repo afaik now." --> am I right?

**Comment by nihilus** [comment]
Yeah seems like it... So it should be closed, right?

**Comment by pombredanne** [comment]
@nihilus you wrote:
> that won't help me much for implementing PDB-support on Linux, will it?

If you do this, please ping when you are there. I can put this to good use in scancode https://github.com/nexB/scancode-toolkit.... At the moment my pdb support is well, just  a plan with a lot of links ;) https://github.com/nexB/scancode-toolkit-contrib/blob/develop/src/compiledcode/pdb.py

**Comment by pombredanne** [comment]
OK, I actually checked the other day and this code does not build at all. 
@msftgits @JimRadigan @GreatKeeper @AndrewPardoe any reason why you could not get  some minimally working/compiling code in? The code looks rather useless in this state.

**Comment by jonwil** [comment]
I think the easiest/fastest way to help out the community here would be to publish a version of cvdump.exe that included a full complete source-level PDB file (with coverage over the entire codebase of cvdump). That would allow people to single-step through those parts of cvdump where the source is in the repo and then see exactly how cvdump decodes specific records/data/etc in a given PDB file.

Complete source code for cvdump and its libraries would be nice but I am sure there are valid reasons why some parts of the code can't be published at this time, hence the idea of being able to single step through the code we do have.

**Comment by pombredanne** [comment]
> Complete source code for cvdump and its libraries would be nice but I am sure there are valid reasons why some parts of the code can't be published at this time, hence the idea of being able to single step through the code we do have.

And what could these reasons be?

**Comment by zjturner** [comment]
If you just want something to step through to see how to decode certain records, use `llvm-pdbdump` in the LLVM repo.  It's not quite as complete as `cvdump` (yet), but it handles all the basics.

Alternatively, you don't _need_ to be able to step through something to understand it.  You can just read the code, all the important stuff is there.

**Comment by pombredanne** [comment]
@zjturner thanks! What I am looking for very specifically is something like `dwarfdump` that can run on Linux, macOS and Windows and dump a parsed PDB file so it can used for analysis purpose e.g. something I could plug in my tools the same way I can plug GNU binutils nm, readelf, objdump or dwarfdump here: https://github.com/nexB/scancode-toolkit-contrib/blob/develop/src/compiledcode/dwarf.py

Do you think https://github.com/llvm-mirror/llvm/blob/master/tools/llvm-pdbdump/llvm-pdbdump.cpp has enough and can be built as a standalone CLI for my three OSes? that would be awesome! (eventually having a parseable output such as JSON would be even better, but I can deal with text alright)

**Comment by zjturner** [comment]
@pombredanne Yes, it already compiles on every platform supported by LLVM, which means it can even run on a big endian system.  And it can already dump to YAML (which is a superset of JSON).  It's still being actively developed, so there are definitely still some obscure corners of PDB that aren't covered by this tool.  Anyway, further questions should be directed to llvm-dev mailing list or the LLVM bug tracker.

**Comment by nihilus** [comment]
@zjturner you seem to miss the point... Can it be integrated in 3rd party with ease is what @pombredanne actually asks.

**Comment by zjturner** [comment]
@nihilus Sorry, which point am I missing?  It's a tool.  It runs on all OSes.  It can print YAML.  Is that not enough to constitute "can be integrated in 3rd party with ease"?  If it is not enough, what piece of additional information do you need to know in order to determine if it can be integrated with ease?

**Comment by pombredanne** [comment]
@zjturner this sounds exactly what the doctor ordered. And frankly I could care less for missing of of the darker corners: if I have something that works like 90% of the times I will be a happy camper. I doubt that the missing code here would ever be provided to compile on POSIX anyway. Thanks!

**Comment by moyix** [comment]
For what it's worth I tried the `llvm-pdbutil` approach just now on a collection of PDB files derived by running symchk on all executables on my Windows VM. It's not really ready for prime time yet: `pdb2yaml` refuses to process stripped PDB files (because the TPI stream is empty, which causes parsing to fail), and `pretty` seems to just segfault on most files. I would report these issues, but they seem to have closed account registrations on their bug tracker due to spam :\

So, a working cvdump tool would still be really nice...

**Comment by jrmuizel** [comment]
Can you post links to the broken PDBs? In my experience, problems with llvm-pdbutil tend to be fixed quickly once reported.

**Comment by moyix** [comment]
@jrmuizel Sure, here are two examples. Is there a better place to have this discussion? I imagine Microsoft does not care too much about bugs in llvm-pdbutil :)

* [win32k.pdb](http://panda.moyix.net/~moyix/llvm_pdb_bug/win32k.pdb) – refuses to parse due to empty TPI stream
* [ntkrnlmp.pdb](http://panda.moyix.net/~moyix/llvm_pdb_bug/ntkrnlmp.pdb) – failure due to unknown symbol kind:
```
Unknown symbol kind!
UNREACHABLE executed at /home/moyix/git/llvm-master/lib/ObjectYAML/CodeViewYAMLSymbols.cpp:480!
```

**Comment by zjturner** [comment]
@moyix Use the email address on the llvm bugzilla page to request a new account.  It's faster than you think.

**Comment by jrmuizel** [comment]
https://bugs.llvm.org/show_bug.cgi?id=33427 and https://bugs.llvm.org/show_bug.cgi?id=33428

**Comment by moyix** [comment]
Thanks! I also got myself an account. @zjturner was right, it was quicker than I expected :)

**Comment by jonwil** [comment]
It seems like Microsoft has abandoned this repo and has no plans to publish anything more so it looks like getting the llvm tools up to scratch is the best way to go as far as having code you can compile and use on other platforms and etc.
